### PR TITLE
Lifetime validation: Remove exceptions

### PR DESCRIPTION
**Pull request by iNinja**
# Lifetime validation: Remove exceptions

- [x] You've read the [Contributor Guide](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/dev/Contributing.md) and [Code of Conduct](https://opensource.microsoft.com/codeofconduct/).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [ ] If any gains or losses in performance are possible, you've included benchmarks for your changes. [More info](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/wiki/Benchmarking-in-Wilson)
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

Removed exceptions thrown as part of lifetime validation.

## Description
- Implemented new (internal momentarily) method for lifetime validation which replaces exception throwing with a return of the new LifetimeValidationResult containing the exception information.
- Added new delegate signature for lifetime validation
- Added tests around all the previous cases.
- Updated ambiguous references in documentation to specify the old version of ValidateLifetime.

### CODE REVIEW

File: test/Microsoft.IdentityModel.TestUtils/IdentityComparer.cs [hunk 1]
```diff
@@ -673,6 +673,74 @@ public static bool AreAudienceValidationResultsEqual(object object1, object obje
...
+
+            return AreLifetimeValidationResultsEqual(
+                object1 as LifetimeValidationResult,
+                object2 as LifetimeValidationResult,
+                "LifetimeValidationResult1",
+                "LifetimeValidationResult2",
+                null,
+                context);
+        }
+        internal static bool AreLifetimeValidationResultsEqual(
```
Comment (keegan-caruso) [inline test/Microsoft.IdentityModel.TestUtils/IdentityComparer.cs:691]: nit: should have a new line between methods

File: test/Microsoft.IdentityModel.Tokens.Tests/Validation/LifetimeValidationResultTests.cs [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+
+            TestUtilities.AssertFailIfErrors(context);
+        }
+
+        public static TheoryData<ValidateLifetimeTheoryData> ValidateLifetimeTestCases
+        {
+            get
+            {
+                DateTime now = DateTime.UtcNow;
+                DateTime oneHourFromNow = EpochTime.DateTime(EpochTime.GetIntDate(DateTime.UtcNow + TimeSpan.FromHours(1)));
```
Comment (keegan-caruso) [inline test/Microsoft.IdentityModel.Tokens.Tests/Validation/LifetimeValidationResultTests.cs]: You don't need to change this, but more of a heads up.

There are slightly easier to read Apis for DateTime, for example

```cs
DateTime.UtcNow.AddHours(1);
```

Comment (keegan-caruso) [reply test/Microsoft.IdentityModel.Tokens.Tests/Validation/LifetimeValidationResultTests.cs]: Sorry I'm missing something here,

Why did we go DateTime -> long -> DateTime?

File: test/Microsoft.IdentityModel.Tokens.Tests/Validation/LifetimeValidationResultTests.cs [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+{
+    public class LifetimeValidationResultTests
+    {
+        [Theory, MemberData(nameof(ValidateLifetimeTestCases), DisableDiscoveryEnumeration = true)]
+        public void ValidateLifetime(ValidateLifetimeTheoryData theoryData)
+        {
+            CompareContext context = TestUtilities.WriteHeader($"{this}.LifetimeValidatorTests", theoryData);
+
+            LifetimeValidationResult lifetimeValidationResult = Validators.ValidateLifetime(
+                theoryData.NotBefore,
```
Comment (keegan-caruso) [inline test/Microsoft.IdentityModel.Tokens.Tests/Validation/LifetimeValidationResultTests.cs:20]: I think we are missing a test case for null notbefore?

File: src/Microsoft.IdentityModel.Tokens/Validation/Validators.Lifetime.cs [hunk 4]
```diff
@@ -2,10 +2,30 @@
...
 namespace Microsoft.IdentityModel.Tokens
 {
+    /// <summary>
+    /// Definition for delegate that will validate the lifetime of a <see cref="SecurityToken"/>.
+    /// </summary>
+    /// <param name="notBefore">The 'notBefore' time found in the <see cref="SecurityToken"/>.</param>
+    /// <param name="expires">The 'expiration' time found in the <see cref="SecurityToken"/>.</param>
+    /// <param name="securityToken">The <see cref="SecurityToken"/> that is being validated.</param>
+    /// <param name="validationParameters"><see cref="TokenValidationParameters"/> required for validation.</param>
+    /// <param name="callContext"></param>
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Tokens/Validation/Validators.Lifetime.cs:19]: this is empty

File: src/Microsoft.IdentityModel.Tokens/Validation/Validators.Lifetime.cs [hunk 5]
```diff
@@ -46,5 +66,142 @@ public static void ValidateLifetime(DateTime? notBefore, DateTime? expires, Secu
...
         }
+
+        /// <summary>
+        /// Validates the lifetime of a <see cref="SecurityToken"/>.
+        /// </summary>
+        /// <param name="notBefore">The 'notBefore' time found in the <see cref="SecurityToken"/>.</param>
+        /// <param name="expires">The 'expiration' time found in the <see cref="SecurityToken"/>.</param>
+        /// <param name="securityToken">The <see cref="SecurityToken"/> being validated.</param>
+        /// <param name="validationParameters"><see cref="TokenValidationParameters"/> required for validation.</param>
+        /// <param name="callContext"></param>
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Tokens/Validation/Validators.Lifetime.cs:77]: this is empty

File: src/Microsoft.IdentityModel.Tokens/Validation/LifetimeValidationResult.cs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    internal class LifetimeValidationResult : ValidationResult
+    {
+        private Exception? _exception;
+
+        /// <summary>
+        /// Creates an instance of <see cref="LifetimeValidationResult"/>
+        /// </summary>
+        /// <paramref name="notBefore"/> is the date from which the token that was validated successfully is valid.
+        /// <paramref name="expires"/> is the expiration date for the token that was validated successfully.
+        public LifetimeValidationResult(DateTime? notBefore, DateTime? expires)
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Tokens/Validation/LifetimeValidationResult.cs:21]: I guess these could be null when validated from a delegate?

File: test/Microsoft.IdentityModel.Tokens.Tests/Validation/LifetimeValidationResultTests.cs (hunk 2 above)
Comment (iNinja) [reply test/Microsoft.IdentityModel.Tokens.Tests/Validation/LifetimeValidationResultTests.cs]: These were written by Copilot. I just named the variable and trusted it to come up with a good way to add the time to the DateTime object. I can change to the format you wrote above.

File: src/Microsoft.IdentityModel.Tokens/Validation/Validators.Lifetime.cs (hunk 4 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens/Validation/Validators.Lifetime.cs:19]: Following up on our chat regarding all `CallContext` references.

File: src/Microsoft.IdentityModel.Tokens/Validation/Validators.Lifetime.cs (hunk 5 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens/Validation/Validators.Lifetime.cs:77]: Following up on our chat regarding all `CallContext` references.

File: test/Microsoft.IdentityModel.Tokens.Tests/Validation/LifetimeValidationResultTests.cs (hunk 3 above)
Comment (iNinja) [reply test/Microsoft.IdentityModel.Tokens.Tests/Validation/LifetimeValidationResultTests.cs:20]: Added test case.

File: src/Microsoft.IdentityModel.Tokens/Validation/LifetimeValidationResult.cs (hunk 6 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens/Validation/LifetimeValidationResult.cs:21]: Added a test to highlight the case where these can be null. Mentioned on the signing key PR about how there are flags to disable validations that make these methods return positive under potential null values.

File: test/Microsoft.IdentityModel.Tokens.Tests/Validation/LifetimeValidationResultTests.cs (hunk 2 above)
Comment (iNinja) [reply test/Microsoft.IdentityModel.Tokens.Tests/Validation/LifetimeValidationResultTests.cs]: Code was updated.

File: test/Microsoft.IdentityModel.TestUtils/IdentityComparer.cs (hunk 1 above)
Comment (iNinja) [reply test/Microsoft.IdentityModel.TestUtils/IdentityComparer.cs:691]: Fixed.

**Review by GeoK — APPROVED** [review]

**Review by keegan-caruso — APPROVED** [review]

**Review by FuPingFranco — APPROVED** [review]
